File: src/lib/shipping.ts
import type { StoreConfig } from '@/types';

export const DEFAULT_FREE_SHIPPING_MIN = 50;
export const DEFAULT_SHIPPING_FEE = 4.99;

export function effectiveShippingConfig(config?: Pick<StoreConfig, 'freeShippingMinOrderAmount' | 'shippingFee'> | null) {
  let freeShippingMinOrderAmount = DEFAULT_FREE_SHIPPING_MIN;
  let shippingFee = DEFAULT_SHIPPING_FEE;
  if (config) {
    if (config.freeShippingMinOrderAmount != null && config.freeShippingMinOrderAmount > 0) {
      freeShippingMinOrderAmount = config.freeShippingMinOrderAmount;
    }
    if (config.shippingFee != null && config.shippingFee > 0) {
      shippingFee = config.shippingFee;
    }
  }
  return { freeShippingMinOrderAmount, shippingFee };
}

/** Shipping charged on merchandise subtotal (after coupon at checkout). */
export function computeShippingFee(merchandiseSubtotal: number, config?: Pick<StoreConfig, 'freeShippingMinOrderAmount' | 'shippingFee'> | null): number {
  const { freeShippingMinOrderAmount, shippingFee } = effectiveShippingConfig(config);
  if (merchandiseSubtotal >= freeShippingMinOrderAmount) return 0;
  return shippingFee;
}

export function amountUntilFreeShipping(merchandiseSubtotal: number, freeShippingMinOrderAmount: number): number {
  return Math.max(0, freeShippingMinOrderAmount - merchandiseSubtotal);
}

File: backend/internal/shipping/shipping.go
package shipping

import "shopreturngifts-api/internal/models"

const (
	DefaultFreeShippingMin = 50.0
	DefaultShippingFee     = 4.99
)

// Defaults returns the effective free-shipping threshold and flat shipping fee.
// Zero or negative stored values mean "use default" for that field.
func Defaults(cfg *models.StoreConfig) (minOrder, fee float64) {
	minOrder, fee = DefaultFreeShippingMin, DefaultShippingFee
	if cfg == nil {
		return minOrder, fee
	}
	if cfg.FreeShippingMinOrderAmount > 0 {
		minOrder = cfg.FreeShippingMinOrderAmount
	}
	if cfg.ShippingFee > 0 {
		fee = cfg.ShippingFee
	}
	return minOrder, fee
}

// Fee returns the flat delivery charge on merchandise subtotal (after coupon).
// This amount is added to the order total but is not included in the tax base.
func Fee(merchandiseSubtotal float64, cfg *models.StoreConfig) float64 {
	minOrder, flatFee := Defaults(cfg)
	if merchandiseSubtotal >= minOrder {
		return 0
	}
	return flatFee
}
